numpy/random/bit_generator.pyx

```diff
@@ -0,0 +1,458 @@
+            `uint64`. Strings (`'uint32'`, `'uint64'`) are fine. Note that
+            requesting `uint64` will draw twice as many bits as `uint32` for
+            the same `n_words`. This is a convenience for `BitGenerator`s that
+            express their states as `uint64` arrays.
+        Returns
+        -------
+        state : uint32 or uint64 array, shape=(n_words,)
+        """
+        hash_const = INIT_B
+        out_dtype = np.dtype(dtype)
+        if out_dtype == np.dtype(np.uint32):
+            pass
+        elif out_dtype == np.dtype(np.uint64):
+            n_words *= 2
+        else:
+            raise ValueError("only support uint32 or uint64")
+        state = np.zeros(n_words, dtype=np.uint32)
+        src_cycle = cycle(self.pool)
+        for i_dst in range(n_words):
+            data_val = next(src_cycle)
+            data_val ^= hash_const
+            hash_const *= MULT_B
+            data_val *= hash_const
+            data_val ^= data_val >> XSHIFT
+            state[i_dst] = data_val & MASK32
+        if out_dtype == np.dtype(np.uint64):
+            state = state.view(np.uint64)
+        return state
+
+    def spawn(self, n_children):
+        """ Spawn a number of child `SeedSequence`s by extending the
+        `spawn_key`.
+        Parameters
+        ----------
+        n_children : int
+        Returns
+        -------
+        seqs : list of `SeedSequence`s
+        """
+        seqs = []
+        for i in range(self.n_children_spawned,
+                       self.n_children_spawned + n_children):
+            seqs.append(SeedSequence(
+                self.entropy,
+                program_entropy=self.program_entropy,
+                spawn_key=self.spawn_key + (i,),
+                pool_size=self.pool_size,
+            ))
+        return seqs
+
+cdef class BitGenerator():
+    """
+    BitGenerator(seed=None)
+
+    Abstract Base Class for generic BitGenerators, which provide a stream
+    of random bits based on different algorithms.
+
+    Parameters
+    ----------
+    seed: {None, SeedSequence, int, sequence[int]}, optional
+        A SeedSequence to initialize the BitGenerator. If None, one will be
+        created. If an int or a sequence of ints, it will be used as the
+        entropy for creating a SeedSequence.
+
+    Attributes
+    ----------
+    lock: threading.Lock
+        Lock instance that is shared so that the same bit git generator can
+        be used in multiple Generators without corrupting the state. Code that
+        generates values from a bit generator should hold the bit generator's
+        lock.
+
+    SeeAlso
+    -------
+    `SeedSequence`
+    """
+
+    def __init__(self, seed=None):
+        self.lock = Lock()
+
+        self._ctypes = None
+        self._cffi = None
+
+        cdef const char *name = "BitGenerator"
+        self.capsule = PyCapsule_New(<void *>&self._bitgen, name, NULL)
+        if seed is None:
+            seed = SeedSequence()
```

Comment:
Should this be named seed_seq, since the seed is usually the numeric value used to set the initial state?

Reply:
I approve of this color of bikeshed.

Reply:
+1